pilot/cmd/pilot-agent/status/server.go

```diff
@@ -601,6 +627,24 @@ func (s *Server) handleAppProbe(w http.ResponseWriter, req *http.Request) {
 	w.WriteHeader(response.StatusCode)
 }
 
+func (s *Server) handleAppProbeTCPSocket(w http.ResponseWriter, prober *Prober) {
+	port := prober.TCPSocket.Port.IntValue()
+	timeout := time.Duration(prober.TimeoutSeconds) * time.Second
+
+	// FIXME(phil9909): Currently, the IP table rules still seem to redirect traffic from inside the pod into the envoy
+	// when connection using the external IP of the pod. Therefore we hardcode it to localhost for now.
+	// See https://istio.io/latest/blog/2021/upcoming-networking-changes/
+	// addr := fmt.Sprintf("%s:%d", s.appProbersDestination, port)
+	addr := fmt.Sprintf("localhost:%d", port)
+	conn, err := net.DialTimeout("tcp", addr, timeout)
```

Comment:
FYI @howardjohn

Reply:
When you hit this issue, were you using a custom net.Dialer with source address set? see https://github.com/istio/istio/blob/2d99cd2817b2a939c85806020bd0acc29b261d5e/pilot/cmd/pilot-agent/status/server.go#L228

If not - you should use that and it should work. That is why HTTP works; the iptables rules should apply the same for TCP and HTTP traffic

Reply:
Thanks, I updated the code accordingly. It's working now.

I additionally had to fix the unit tests after the change, as `::6` is not a real loopback address.
This revealed that there are no unit tests for the IPv6 based HTTP probes. I can add some in a follow up PR.